### ISSUE TITLE
[INSTALL] Installing via a proxy regressed in 4.x

### ISSUE THREAD

**Issue by AxelTerizaki** [mapbox/node-sqlite3#1010]
Hello.

This is a follow-up on previous issues involving node-pre-gyp.

I am behind a corporate proxy for this. At home this installed just fine.

However, I don't believe the proxy is at fault here, because installing version 3.1.13 works flawlessly

I started a brand new app to test this out, with `npm init`

```
D:\perso\testsqlite>npm install sqlite3@3.1.13

> sqlite3@3.1.13 install D:\perso\testsqlite\node_modules\sqlite3
> node-pre-gyp install --fallback-to-build

[sqlite3] Success: "D:\perso\testsqlite\node_modules\sqlite3\lib\binding\node-v5
7-win32-ia32\node_sqlite3.node" is installed via remote
npm notice created a lockfile as package-lock.json. You should commit this file.

npm WARN testsqlite@1.0.0 No description
npm WARN testsqlite@1.0.0 No repository field.

+ sqlite3@3.1.13
added 114 packages in 20.183s
```

With 4.0.1 it fails : 
```
D:\perso\testsqlite>npm install sqlite3@4.0.1

> sqlite3@4.0.1 install D:\perso\testsqlite\node_modules\sqlite3
> node-pre-gyp install --fallback-to-build

node-pre-gyp WARN Tried to download(501): https://mapbox-node-binary.s3.amazonaw
s.com/sqlite3/v4.0.1/node-v57-win32-ia32.tar.gz
node-pre-gyp WARN Pre-built binaries not found for sqlite3@4.0.1 and node@8.11.3
 (node-v57 ABI, unknown) (falling back to source compile with node-gyp)
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "D:\apps\python3\python.EXE",
 you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (d:\Apps\node\node_modules\npm\n
ode_modules\node-gyp\lib\configure.js:483:19)
gyp ERR! stack     at PythonFinder.<anonymous> (d:\Apps\node\node_modules\npm\no
de_modules\node-gyp\lib\configure.js:508:16)
gyp ERR! stack     at d:\Apps\node\node_modules\npm\node_modules\graceful-fs\pol
yfills.js:284:29
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:152:21)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "d:\\apps\\node\\node.exe" "d:\\Apps\\node\\node_modules\\npm\\
node_modules\\node-gyp\\bin\\node-gyp.js" "configure" "--fallback-to-build" "--m
odule=D:\\perso\\testsqlite\\node_modules\\sqlite3\\lib\\binding\\node-v57-win32
-ia32\\node_sqlite3.node" "--module_name=node_sqlite3" "--module_path=D:\\perso\
\testsqlite\\node_modules\\sqlite3\\lib\\binding\\node-v57-win32-ia32" "--napi_v
ersion=3" "--node_abi_napi=napi"
gyp ERR! cwd D:\perso\testsqlite\node_modules\sqlite3
gyp ERR! node -v v8.11.3
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
node-pre-gyp ERR! build error
node-pre-gyp ERR! stack Error: Failed to execute 'd:\apps\node\node.exe d:\Apps\
node\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js configure --fallback
-to-build --module=D:\perso\testsqlite\node_modules\sqlite3\lib\binding\node-v57
-win32-ia32\node_sqlite3.node --module_name=node_sqlite3 --module_path=D:\perso\
testsqlite\node_modules\sqlite3\lib\binding\node-v57-win32-ia32 --napi_version=3
 --node_abi_napi=napi' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (D:\perso\testsqlite\nod
e_modules\node-pre-gyp\lib\util\compile.js:83:29)
node-pre-gyp ERR! stack     at emitTwo (events.js:126:13)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:925:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/chi
ld_process.js:209:5)
node-pre-gyp ERR! System Windows_NT 6.1.7601
node-pre-gyp ERR! command "d:\\apps\\node\\node.exe" "D:\\perso\\testsqlite\\nod
e_modules\\node-pre-gyp\\bin\\node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd D:\perso\testsqlite\node_modules\sqlite3
node-pre-gyp ERR! node -v v8.11.3
node-pre-gyp ERR! node-pre-gyp -v v0.10.2
node-pre-gyp ERR! not ok
Failed to execute 'd:\apps\node\node.exe d:\Apps\node\node_modules\npm\node_modu
les\node-gyp\bin\node-gyp.js configure --fallback-to-build --module=D:\perso\tes
tsqlite\node_modules\sqlite3\lib\binding\node-v57-win32-ia32\node_sqlite3.node -
-module_name=node_sqlite3 --module_path=D:\perso\testsqlite\node_modules\sqlite3
\lib\binding\node-v57-win32-ia32 --napi_version=3 --node_abi_napi=napi' (1)
npm WARN testsqlite@1.0.0 No description
npm WARN testsqlite@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! sqlite3@4.0.1 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the sqlite3@4.0.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional log
ging output above.
```

Any help would be appreciated. :)

**Comment by springmeyer** [comment]
The issue may be with https://github.com/mapbox/node-pre-gyp/pull/350. Basically node-pre-gyp does the download and recently node-sqlite3 was updated to the latest node-pre-gyp that includes the change to move from request to needle (https://github.com/mapbox/node-pre-gyp/pull/350). Perhaps either:

  - the node-pre-gyp passing of proxy info to needle is incorrect, or
  - needle itself is buggy

How are you configuring your proxy?

**Comment by springmeyer** [comment]
hmmm, yeah https://github.com/tomas/needle#more-advanced-proxy-support and https://github.com/tomas/needle/issues/176 seem to indicate the this likely regressed due to the move to `needle`. I don't have access to a proxy to debug this - would you be interested in trying to add an agent to node-pre-gyp (and providing a PR) to get this working?

**Comment by AxelTerizaki** [comment]
Thanks for your quick reply :)

I'm configuring it via the environment variable http_proxy and https_proxy as well as npm's config setting https-proxy and proxy.

What do you mean by adding an agent to node-pre-gyp? I'd like to help to get this sorted out but not sure I know what to do.

**Comment by springmeyer** [comment]
> What do you mean by adding an agent to node-pre-gyp
 
@AxelTerizaki read https://github.com/tomas/needle#more-advanced-proxy-support. Doing that would require a code change at https://github.com/mapbox/node-pre-gyp/blob/master/lib/install.js#L31-L65

**Comment by AxelTerizaki** [comment]
I'll see if I can try something out tomorrow, but I admit I'm a little unfamiliar with the code here and how to efficiently test it out without making mistakes.

**Comment by springmeyer** [comment]
#1012 also appears to be proxy related, so closed that one and renamed the title of this one, to help us discuss in one place. Dropping the proxy config (from #1012) that stopped working here:

```
https_proxy=http://xxxxxx.yyy.dom:180
http_proxy=http://xxxxxx.yyy.dom:180
no_proxy=localhost,127.0.0.1,.yyy.dom
```

I think proxy support regressed in node-sqlite3 4.0.0 because that is the version that saw an upgrade to node-pre-gyp 0.9.0 (https://github.com/mapbox/node-sqlite3/blob/master/CHANGELOG.md#400). And node-pre-gyp 0.9.0 is when `needle` replaced `request` (https://github.com/mapbox/node-pre-gyp/blob/master/CHANGELOG.md#090) and likely when proxy support regressed per https://github.com/mapbox/node-pre-gyp/issues/400

**Comment by springmeyer** [comment]
So, anyone hitting an install error who is behind a proxy, these are your options:

 - downgrade to node-sqlite3@3.1.13
 - avoid using a proxy
 - try to help add working proxy support to node-pre-gyp/needle
 - hack the node-pre-gyp code to add back request

**Comment by AxelTerizaki** [comment]
Tried using the `tunnel` module but didn't work out :+1: 
```
node-pre-gyp WARN Pre-built binaries not installable for sqlite3@4.0.1 and node 8.11.3 (node-v57 ABI, unknown) (falling back to source compile with node-gyp)
node-pre-gyp WARN Hit error tunneling socket could not be established, statusCode=403
```

here's what I modified in `install.js` (xxx and yyy are obviously my corporate proxy IP and port) :

```
        var tunnel = require('tunnel');
	var myAgent = tunnel.httpOverHttp({
  		proxy: { host: 'xxx', port: yyy}
	});
    var requestOpts = {
        uri: uri.replace('+','%2B'),
        headers: {
          'User-Agent': 'node-pre-gyp (v' + npgVersion + ', ' + envVersionInfo + ')'
        },
        follow_max: 10,
	agent: myAgent
    };
```

Not too sure if I did something wrong or not here, but not sure if the tunnel thing is the good way to go.

**Comment by springmeyer** [comment]
> Tried using the tunnel module but didn't work out

That's do bad. Thanks for trying! Maybe try asking a question about how to set up over at https://github.com/tomas/needle/issues?

**Comment by springmeyer** [comment]
This should be fixed in `master`. Can you test to confirm? To test do:

```
git clone https://github.com/mapbox/node-sqlite3.git
cd node-sqlite3
npm install
```

**Comment by AxelTerizaki** [comment]
I can confirm it works fine now from being a proxy : 

```$ npm install

> sqlite3@4.0.1 install D:\perso\node-sqlite3
> node-pre-gyp install --fallback-to-build

node-pre-gyp WARN Using request for node-pre-gyp https download
[sqlite3] Success: "D:\perso\node-sqlite3\lib\binding\node-v57-win32-ia32\node_sqlite3.node" is installed via remote
npm notice created a lockfile as package-lock.json. You should commit this file.
added 235 packages in 125.748s
```

Thank you for fixing this !

**Comment by springmeyer** [comment]
Great, this is now fixed in sqlite3@4.0.2, which is now released.
